python/openai/openai_frontend/engine/utils/triton.py

```diff
@@ -0,0 +1,175 @@
+    if request.temperature is not None:
+        inputs["temperature"] = np.float32([[request.temperature]])
+    # FIXME: TRT-LLM doesn't currently support runtime changes of 'echo' and it
+    # is configured at model load time, so we don't handle it here for now.
+    return model.create_request(inputs=inputs)
+
+
+def _construct_string_from_pointer(pointer: int, size: int) -> str:
+    """Constructs a Python string from a C pointer and size."""
+
+    # Create a ctypes string buffer
+    string_buffer = ctypes.create_string_buffer(size + 1)  # +1 for null terminator
+
+    # Copy the data from the pointer to the buffer
+    ctypes.memmove(string_buffer, pointer, size)
+
+    # Convert the buffer to a Python string
+    return string_buffer.value.decode("utf-8")  # Adjust encoding if needed
+
+
+def _get_volume(shape: Iterable[int]) -> int:
+    volume = 1
+    for dim in shape:
+        volume *= dim
+
+    return volume
+
+
+def _to_string(tensor: tritonserver.Tensor) -> str:
+    # FIXME: This could be a bit more robust by reading byte size from first
+    # 4 bytes and then just reading the first string, rather than assuming
+    # single string, assuming it's of similar performance to do so.
+
+    # The following optimization to read string directly from buffer assumes
+    # there is only a single string, so enforce it to avoid obscure errors.
+    volume = _get_volume(tensor.shape)
+    if volume != 1:
+        raise Exception(
+            f"Expected to find 1 string in the output, found {volume} instead."
+        )
+    if tensor.size < 4:
+        raise Exception(
+            f"Expected string buffer to contain its serialized byte size, but found size of {tensor.size}."
+        )
+
+    # NOTE: +/- 4 accounts for serialized byte string length in first 4 bytes of buffer
+    return _construct_string_from_pointer(tensor.data_ptr + 4, tensor.size - 4)
+
+
+# TODO: Use tritonserver.InferenceResponse when support is published
+def _get_output(response: tritonserver._api._response.InferenceResponse) -> str:
+    if "text_output" in response.outputs:
+        tensor = response.outputs["text_output"]
+
+        # Alternative method, creates the same string, but goes through
+        # deserialization, numpy, and dlpack overhead:
+        # return tensor.to_bytes_array()[0].decode("utf-8")
+
+        # Optimized method
+        return _to_string(tensor)
+
+    return ""
+
+
+def _validate_triton_responses_non_streaming(
+    responses: List[tritonserver._api._response.InferenceResponse],
+):
+    num_responses = len(responses)
+    if num_responses == 1 and responses[0].final != True:
+        raise Exception("Unexpected internal error with incorrect response flags")
+    if num_responses == 2 and responses[-1].final != True:
+        raise Exception("Unexpected internal error with incorrect response flags")
+    if num_responses > 2:
+        raise Exception(f"Unexpected number of responses: {num_responses}, expected 1.")
```

Comment:
Yeah agreed the separation of the utils/helpers definitely got a little blurry/messy - a restructure like this makes sense.

I would like to further restructure later a few things as well.

I think eventually having imports in a user-facing app looking something like this would look nice:
```python
from tritonserver.core import Server
from tritonserver.frontends.kserve import KServeGrpc
from tritonserver.frontends.openai import FastApiFrontend
from tritonserver.frontends.openai.engine import TritonLLMEngine
```

But will consider restructuring in general as a follow-up PR focused just on structure after some discussion.